transformer_engine/pytorch/optimizers/fused_adam.py

```diff
@@ -232,68 +198,80 @@ def step(self, closure=None, grad_scaler=None):
-                        scales.append(scale)
-                        amaxes.append(amax)
-                        scale_invs.append(scale_inv)
-                        p_main_of_fp8_model.append(p.data)
-                        g_main_of_fp8_model.append(p.grad.data)
-                        m_of_fp8_model.append(state["exp_avg"])
-                        v_of_fp8_model.append(state["exp_avg_sq"])
-                    elif p_extra.dtype in [torch.float16, torch.bfloat16]:
-                        p_f16_model.append(p_extra.data)
-                        p_main_of_f16_model.append(p.data)
-                        g_main_of_f16_model.append(p.grad.data)
-                        m_of_f16_model.append(state["exp_avg"])
-                        v_of_f16_model.append(state["exp_avg_sq"])
+                    # Master weights
+                    if self.master_weights and p.dtype != torch.float32:
+                        # model weights can be fp32/bf16/fp16/fp8
+                        # If it's fp32, it has no corresponding master weights
+                        state["master_param"] = self.master_weights[master_param_idx]
+                        master_param_idx += 1
+                        assert (
+                            state["master_param"].shape == p.shape
+                        ), "Master weights shape must match model weights shape"
                     else:
-                        raise RuntimeError(
-                            "FusedAdam only support model weights in fp16/bf16 and fp8 when"
-                            " fuse_dtype_casting is True."
-                        )
+                        state["master_param"] = None
+
+                p_master = state["master_param"]
+                p_grad = p.grad
+
+                if self.master_weights and p_master is not None and p_master.grad is not None:
+                    p_grad = p_master.grad
 
+                if p_grad is None:
+                    continue
+                if p_grad.data.is_sparse:
+                    raise RuntimeError("FusedAdam does not support sparse gradients.")
+
+                if isinstance(p, Float8Tensor):
+                    out_dtype = p._fp8_dtype
+                    p_fp8_model.append(p._data.data)
+                    scale, amax, scale_inv = get_fp8_meta(p)
+                    scales.append(scale)
+                    amaxes.append(amax)
+                    scale_invs.append(scale_inv)
+                    if self.master_weights:
+                        p_main_of_fp8_model.append(p_master.data)
+                    g_of_fp8_model.append(p_grad.data)
+                    m_of_fp8_model.append(state["exp_avg"])
+                    v_of_fp8_model.append(state["exp_avg_sq"])
+                elif p.dtype in [torch.float16, torch.bfloat16]:
+                    has_fp16 = True if p.dtype == torch.float16 else False
+                    has_bf16 = True if p.dtype == torch.bfloat16 else False
```

Comment:
These are reset every loop iteration:
```suggestion
                    has_fp16 = has_fp16 or p.dtype == torch.float16
                    has_bf16 = has_bf16 or p.dtype == torch.bfloat16
```